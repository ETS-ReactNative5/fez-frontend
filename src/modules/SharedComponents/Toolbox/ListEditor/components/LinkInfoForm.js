import React, { useState, useRef, useCallback } from 'react';
import PropTypes from 'prop-types';

import TextField from '@material-ui/core/TextField';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';

export const LinkInfoForm = ({ disabled, locale, errorText, onAdd }) => {
    const [linkAndDescription, setLinkAndDescription] = useState({ key: null, value: null });
    const linkInput = useRef(null);
    const descriptionInput = useRef(null);

    const handleChange = (event) => {
        const { name, value } = event.target;
        setLinkAndDescription({
            ...linkAndDescription,
            [name]: value,
        });
    };

    const resetForm = () => {
        setLinkAndDescription({
            key: null,
            value: null,
        });
        linkInput.current.value = null;
        descriptionInput.current.value = null;
    };

    const addItem = useCallback(
        (event) => {
            // add item if user hits 'enter' key on input field
            if (
                disabled ||
                !linkAndDescription.key ||
                !linkAndDescription.value ||
                (event && event.key && event.key !== 'Enter')
            ) {
                return;
            }
            // pass on the selected item
            onAdd(linkAndDescription);
            resetForm();
            // move focus to name as published text field after item was added
            /* istanbul if ignore */
            descriptionInput.current.focus();
            // linkInput.current.focus();
        },
        [disabled, linkAndDescription, onAdd]
    );

    const {
        linkInputFieldLabel,
        linkInputFieldHint,
        descriptionInputFieldLabel,
        descriptionInputFieldHint,
        addButtonLabel,
        id,
    } = locale;

    return (
        <Grid container spacing={16} display="row" alignItems="center">
            <Grid item style={{ flexGrow: 1 }} xs={12} sm={6} md={5}>
                <TextField
                    fullWidth
                    name="key"
                    id={(!!id && id) || ''}
                    label={linkInputFieldLabel}
                    placeholder={linkInputFieldHint}
                    onChange={handleChange}
                    onKeyPress={addItem}
                    error={!!errorText}
                    disabled={disabled}
                    inputProps={{
                        ref: linkInput,
                    }}
                />
            </Grid>
            <Grid item xs={12} sm={6} md={5}>
                <TextField
                    fullWidth
                    name="value"
                    id={(!!id && id) || ''}
                    label={descriptionInputFieldLabel}
                    placeholder={descriptionInputFieldHint}
                    onChange={handleChange}
                    onKeyPress={addItem}
                    error={!!errorText}
                    disabled={disabled}
                    inputProps={{
                        ref: descriptionInput,
                    }}
                />
            </Grid>
            <Grid item xs={12} md={2}>
                <Button
                    fullWidth
                    id="add-items"
                    color="primary"
                    variant="contained"
                    children={addButtonLabel}
                    disabled={disabled}
                    onClick={addItem}
                />
            </Grid>
        </Grid>
    );
};

LinkInfoForm.propTypes = {
    onAdd: PropTypes.func.isRequired,
    locale: PropTypes.object,
    disabled: PropTypes.bool,
    errorText: PropTypes.string,
};

export default React.memo(LinkInfoForm);
