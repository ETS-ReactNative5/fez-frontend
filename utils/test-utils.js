import React from 'react';
import PropTypes from 'prop-types';
import { render } from 'react-testing-library';
import {Router} from 'react-router-dom';
import {mui1theme} from 'config/theme';
import {Provider} from 'react-redux';
import MuiThemeProvider from '@material-ui/core/styles/MuiThemeProvider';
import MuiPickersUtilsProvider from 'material-ui-pickers/MuiPickersUtilsProvider';
import MomentUtils from '@date-io/moment';

import {getStore} from '../src/config/store';
import Immutable from 'immutable';
import {createMemoryHistory} from 'history';

const domTestingLib = require('dom-testing-library');
const reactTestingLib = require('react-testing-library');

const { queryHelpers } = domTestingLib;

const queryByTestId = queryHelpers.queryByAttribute.bind(
    null,
    'id'
);

const queryAllByTestId = queryHelpers.queryAllByAttribute.bind(
    null,
    'id'
);

function getAllByTestId(container, id, ...rest) {
    const els = queryAllByTestId(container, id, ...rest);
    if (!els.length) {
        throw queryHelpers.getElementError(
            `Unable to find an element by: [id="${id}"]`,
            container
        );
    }
    return els;
}

function getByTestId(...args) {
    return queryHelpers.firstResultOrNull(getAllByTestId, ...args);
}


const AllTheProviders = (props) => {
    return (
        <MuiThemeProvider theme={mui1theme}>
            <MuiPickersUtilsProvider utils={MomentUtils}>
                {props.children}
            </MuiPickersUtilsProvider>
        </MuiThemeProvider>
    );
};

AllTheProviders.propTypes = {
    children: PropTypes.node
};

export const rtlRender = (ui, options) =>
    render(ui, { wrapper: AllTheProviders, ...options });

export const renderWithRouter = (ui, {
    route = '/',
    history = createMemoryHistory({initialEntries: [route]})
} = {}) => {
    return {
        ...rtlRender(<Router history={history}>{ui}</Router>),
        history,
    };
};

export const renderWithRedux = ({initialState}) => (render) => {
    return {
        ...render,
        store: getStore({initialState, history: render.history})
    };
};

export const withRouter = ({
    route = '/',
    history = createMemoryHistory({initialEntries: [route]})} = {}
) => (WrappedComponent) => {
    return (<Router history={history}>{WrappedComponent}</Router>);
};

export const withRedux = (initialState = Immutable.Map()) => (WrappedComponent) => {
    return (<Provider store={getStore(initialState)}>{WrappedComponent}</Provider>);
};

module.exports = {
    ...domTestingLib,
    ...reactTestingLib,
    queryByTestId,
    queryAllByTestId,
    getAllByTestId,
    getByTestId,
    rtlRender,
    renderWithRouter,
    renderWithRedux,
    withRedux,
    withRouter
};
