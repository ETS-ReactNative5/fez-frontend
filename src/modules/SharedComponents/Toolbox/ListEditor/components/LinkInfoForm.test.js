import React from 'react';
import LinkInfoForm from './LinkInfoForm';
import { rtlRender, fireEvent } from 'test-utils';

const setup = (testProps = {}) => {
    const props = {
        onAdd: jest.fn(),
        locale: {
            linkInputFieldLabel: 'Link',
            linkInputFieldHint: 'Enter link',
            descriptionInputFieldLabel: 'Description',
            descriptionInputFieldHint: 'Enter description',
            addButtonLabel: 'Add link',
            editButtonLabel: 'Update link',
        },
        ...testProps,
    };

    return rtlRender(<LinkInfoForm {...props} />);
};

describe('LinkInfoForm', () => {
    it('should render link info form with two inputs and button disabled', () => {
        const { getByTestId } = setup();
        expect(getByTestId('link-info-link')).toBeVisible();
        expect(getByTestId('link-info-description')).toBeVisible();
        expect(getByTestId('add-items')).toHaveAttribute('disabled', '');
    });

    it('should enable "Add link" button if valid link is entered', () => {
        const { getByTestId } = setup();

        fireEvent.change(getByTestId('link-info-link'), { target: { value: 'http://test.com' } });
        expect(getByTestId('add-items')).not.toHaveAttribute('disabled', '');
    });

    it('should display error message if link is not valid', () => {
        const { getByTestId, getByText } = setup();

        fireEvent.change(getByTestId('link-info-link'), { target: { value: 'test.com' } });
        expect(getByText('URL is not valid')).toBeVisible();
        expect(getByTestId('add-items')).toHaveAttribute('disabled', '');
    });

    it('should add link info and reset form if link is valid and "Enter" is pressed', () => {
        const onAddFn = jest.fn();
        const { getByTestId } = setup({ onAdd: onAddFn });

        fireEvent.change(getByTestId('link-info-link'), { target: { value: 'http://test.com' } });
        fireEvent.change(getByTestId('link-info-description'), { target: { value: 'test description' } });
        fireEvent.keyDown(getByTestId('link-info-description'), { key: 'Enter', code: 13 });

        expect(onAddFn).toHaveBeenCalledWith({ key: 'http://test.com', value: 'test description' });
        expect(getByTestId('link-info-link')).toHaveAttribute('value', '');
        expect(getByTestId('link-info-description')).toHaveAttribute('value', '');
    });

    it('should not submit link info form if link is not valid and "Enter" is pressed', () => {
        const onAddFn = jest.fn();
        const { getByTestId } = setup({ onAdd: onAddFn });

        fireEvent.change(getByTestId('link-info-link'), { target: { value: 'test.com' } });
        fireEvent.change(getByTestId('link-info-description'), { target: { value: 'test description' } });
        fireEvent.keyDown(getByTestId('link-info-description'), { key: 'Enter', code: 13 });

        expect(onAddFn).not.toHaveBeenCalled();
    });

    it('should not submit link info form if link is valid and "Esc" is pressed', () => {
        const onAddFn = jest.fn();
        const { getByTestId } = setup({ onAdd: onAddFn });

        fireEvent.change(getByTestId('link-info-link'), { target: { value: 'http://test.com' } });
        fireEvent.change(getByTestId('link-info-description'), { target: { value: 'test description' } });
        fireEvent.keyDown(getByTestId('link-info-description'), { key: 'Esc', code: 27 });

        expect(onAddFn).not.toHaveBeenCalled();
    });

    it('should load link info form with give link info values', () => {
        const onAddFn = jest.fn();
        const { getByTestId } = setup({
            onAdd: onAddFn,
            itemSelectedToEdit: { key: 'http://test.com', value: 'test description' },
        });

        expect(getByTestId('link-info-link')).toHaveAttribute('value', 'http://test.com');
        expect(getByTestId('link-info-description')).toHaveAttribute('value', 'test description');
    });
});
