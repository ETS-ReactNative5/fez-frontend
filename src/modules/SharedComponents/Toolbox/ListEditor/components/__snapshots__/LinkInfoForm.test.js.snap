// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkInfoForm component should render the default view 1`] = `
<WithStyles(Grid)
  alignItems="center"
  container={true}
  display="row"
  spacing={16}
>
  <WithStyles(Grid)
    item={true}
    md={5}
    sm={6}
    style={
      Object {
        "flexGrow": 1,
      }
    }
    xs={12}
  >
    <TextField
      error={false}
      fullWidth={true}
      id=""
      inputProps={
        Object {
          "ref": Object {
            "current": null,
          },
        }
      }
      name="key"
      onChange={[Function]}
      onKeyPress={[Function]}
      required={false}
      select={false}
      variant="standard"
    />
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    md={5}
    sm={6}
    xs={12}
  >
    <TextField
      error={false}
      fullWidth={true}
      id=""
      inputProps={
        Object {
          "ref": Object {
            "current": null,
          },
        }
      }
      name="value"
      onChange={[Function]}
      onKeyPress={[Function]}
      required={false}
      select={false}
      variant="standard"
    />
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    md={2}
    xs={12}
  >
    <WithStyles(Button)
      color="primary"
      fullWidth={true}
      id="add-items"
      onClick={[Function]}
      variant="contained"
    />
  </WithStyles(Grid)>
</WithStyles(Grid)>
`;
